Sources/Common/Communication/EventBusHandler.swift

```diff
@@ -0,0 +1,113 @@
+    ///   - eventType: The event type to observe.
+    ///   - action: The action to execute when the event is observed.
+    public func addObserver<E: EventRepresentable>(_ eventType: E.Type, action: @escaping (E) -> Void) {
+        logger.debug("EventBusHandler: Adding observer for event type - \(eventType)")
+
+        let adaptedAction: (AnyEventRepresentable) -> Void = { event in
+            if let specificEvent = event as? E {
+                action(specificEvent)
+            } else {
+                self.logger.debug("Error: Event type did not match")
+            }
+        }
+
+        Task {
+            await eventBus.addObserver(eventType.key, action: adaptedAction)
+            await replayEvents(forType: eventType)
+        }
+    }
+
+    /// Removes an observer for a specific event type.
+    /// - Parameter eventType: The event type for which to remove the observer.
+    public func removeObserver<E: EventRepresentable>(for eventType: E.Type) {
+        Task { await eventBus.removeObserver(for: E.key) }
+    }
+
+    /// Replays events of a specific type to any new observers, ensuring they receive past events.
+    /// - Parameter eventType: The event type for which to replay events.
+    private func replayEvents<E: EventRepresentable>(forType eventType: E.Type) async {
+        let key = eventType.key
+        let storedEvents = await memoryStorage.eventsForKey(key)
+
+        for event in storedEvents {
+            if let specificEvent = event as? E {
+                logger.debug("EventBusHandler: Replaying event type - \(specificEvent)")
+                let isSent = await eventBus.post(specificEvent)
+                if isSent {
+                    await removeFromStorage(specificEvent)
+                }
+            }
+        }
+    }
+
+    /// Posts an event to the EventBus and stores it if there are no observers.
+    /// - Parameter event: The event to post.
+    public func postEvent<E: EventRepresentable>(_ event: E) {
+        logger.debug("EventBusHandler: Posting event - \(event)")
+        Task {
+            let hasObservers = await eventBus.post(event)
+            await memoryStorage.appendEvent(event)
+            if !hasObservers {
+                logger.debug("EventBusHandler: Storing event in memory - \(event)")
+                await storeEvent(event)
+            }
+        }
+    }
+
+    /// Stores an event in persistent storage.
+    /// - Parameter event: The event to store.
+    private func storeEvent<E: EventRepresentable>(_ event: E) async {
+        do {
+            try await eventStorage.store(event: event)
+        } catch {
+            logger.debug("Error storing event: \(error)")
+        }
+    }
+
+    /// Removes an event from persistent storage.
+    /// - Parameter event: The event to remove.
+    public func removeFromStorage<E: EventRepresentable>(_ event: E) async {
+        await eventStorage.remove(ofType: event.key, withStorageId: event.storageId)
```

Comment:
when do we clean up the in memory storage? so far I see we are only appending to it. Which at one point will need a clean up

Reply:
I want to think we don't clean up the memory, it would be cleaned on SDK/App restart, and we could just limit the number of items that get added to the memory. 

Even though there is only a limited number of events that gets sent via eventbus in SDK lifecycle, if an anomaly happen and users starts bombarding with those events we only keep X amount of latest events to be replayed to new consumers.